src/node/tendermint_processes.ml

```diff
@@ -0,0 +1,367 @@
+  let _round = consensus_state.round in
+  let do_something (consensus_state : consensus_state) (found_set : MySet.t) =
+    let value, _ = MySet.choose found_set in
+    consensus_state.step <- Prevote;
+    if
+      is_valid global_state value
+      && (consensus_state.locked_round = -1
+         || consensus_state.locked_value = value)
+    then
+      Broadcast (PrevoteOP (height, round, repr_of_value value))
+    else
+      Broadcast (PrevoteOP (height, round, nil)) in
+  let found_set =
+    select_proposal_process_round_matching_proposer msg_log consensus_state
+      global_state in
+  let extract_valid_set (found_set : MySet.t) =
+    MySet.filter (fun (_, r) -> r = round) found_set in
+  let valid_set = extract_valid_set found_set in
+  if valid_set = MySet.empty || step <> Proposal then
+    None
+  else
+    Some (do_something consensus_state valid_set)
+
+let precommit_failed_previous_round (_height : height) (_round : round)
+    (consensus_state : consensus_state) (msg_log : input_log) _dlog global_state
+    =
+  let height = consensus_state.height in
+  let step = consensus_state.step in
+  let round = consensus_state.round in
+  (* Returns list of (repr_value, round) with filtered round *)
+  let extract_common_set (left_set : MySet.t) (right_set : MySet.t) : MySet.t =
+    let filtered_left_set =
+      MySet.filter (fun (_, r) -> r >= 0 && r < round) left_set in
+    (* (v, r)*)
+    let filtered_right_set =
+      MySet.filter (fun (_, r) -> r >= 0 && r < round) right_set in
+    (* (rv, r)*)
+    MySet.filter
+      (fun (v, r) -> MySet.mem (repr_of_value v, r) filtered_right_set)
+      filtered_left_set in
+  let do_something (consensus_state : consensus_state) (common_set : MySet.t) =
+    let value, valid_round = MySet.choose common_set in
+    consensus_state.step <- Prevote;
+    if
+      is_valid global_state value
+      && (consensus_state.locked_round <= valid_round
+         || consensus_state.locked_value = value)
+    then
+      Broadcast (PrevoteOP (height, round, repr_of_value value))
+    else
+      Broadcast (PrevoteOP (height, round, nil)) in
+  let left_set =
+    select_proposal_valid_round_matching_proposer msg_log consensus_state
+      global_state in
+  let right_set = count_prevotes msg_log consensus_state global_state in
+  let common_set = extract_common_set left_set right_set in
+  if
+    left_set = MySet.empty
+    || right_set = MySet.empty
+    || step <> Proposal
+    || common_set = MySet.empty
+  then
+    None
+  else
+    Some (do_something consensus_state common_set)
+
+let prepare_default_precommit_prevote_phase (height : height) (round : round)
+    (consensus_state : consensus_state) (msg_log : input_log) _dlog global_state
+    =
+  let take_all_prevotes = function
+    | PrevoteContent content when content.process_round = consensus_state.round
+      ->
+      Some ((nil, content.process_round), get_weight global_state content.sender)
+    | PrevoteContent _ -> None
```

Comment:
I don't get the question. `count_prevotes` takes a `prevote_selection` that is defined here where what matters is that round match.

Reply:
I didn't know the `when` syntax right before, so I was asking if this pattern was useful because I thought line222 would be the same as line219.
My bad.